source/server/admin/logs_handler.cc

```diff
@@ -56,56 +61,106 @@ bool LogsHandler::changeLogLevel(const Http::Utility::QueryParams& params) {
+    // not common to call this function at a high rate.
+    absl::flat_hash_map<absl::string_view, absl::string_view> name_levels;
+    std::vector<absl::string_view> pairs = absl::StrSplit(value, ',', absl::SkipWhitespace());
+    for (const auto& name_value : pairs) {
+      std::pair<absl::string_view, absl::string_view> name_value_pair =
+          absl::StrSplit(name_value, absl::MaxSplits(':', 1), absl::SkipWhitespace());
+      auto [name, value] = name_value_pair;
+      if (name.empty() || value.empty()) {
+        continue;
+      }
+
+      name_levels.insert({name, value});
+    }
+
+    return changeLogLevels(name_levels);
+  } else {
+    // Change particular log level by name.
+    return changeLogLevelByName(key, value);
+  }
+}
 
-  // First see if the level is valid.
-  size_t level_to_use = std::numeric_limits<size_t>::max();
+static absl::optional<spdlog::level::level_enum> parseLevel(absl::string_view level) {
   for (size_t i = 0; i < ARRAY_SIZE(spdlog::level::level_string_views); i++) {
-    if (level == spdlog::level::level_string_views[i]) {
-      level_to_use = i;
-      break;
+    spdlog::string_view_t spd_log_level{spdlog::level::level_string_views[i]};
+    if (level == absl::string_view{spd_log_level.data(), spd_log_level.size()}) {
+      return static_cast<spdlog::level::level_enum>(i);
     }
   }
 
-  if (level_to_use == std::numeric_limits<size_t>::max()) {
+  return absl::nullopt;
+}
+
+bool LogsHandler::changeAllLogLevels(absl::string_view level) {
+  auto level_to_use = parseLevel(level);
+  if (!level_to_use.has_value()) {
+    ENVOY_LOG(error, "invalid log level specified: level='{}'", level);
     return false;
   }
 
   if (!Logger::Context::useFancyLogger()) {
-    // Now either change all levels or a single level.
-    if (name == "level") {
-      ENVOY_LOG(debug, "change all log levels: level='{}'", level);
-      for (Logger::Logger& logger : Logger::Registry::loggers()) {
-        logger.setLevel(static_cast<spdlog::level::level_enum>(level_to_use));
-      }
-    } else {
-      ENVOY_LOG(debug, "change log level: name='{}' level='{}'", name, level);
-      Logger::Logger* logger_to_change = nullptr;
-      for (Logger::Logger& logger : Logger::Registry::loggers()) {
-        if (logger.name() == name) {
-          logger_to_change = &logger;
-          break;
-        }
-      }
+    ENVOY_LOG(debug, "change all log levels: level='{}'", level);
+    for (Logger::Logger& logger : Logger::Registry::loggers()) {
+      logger.setLevel(*level_to_use);
+    }
+  } else {
+    // Level setting with Fancy Logger.
+    FANCY_LOG(info, "change all log levels: level='{}'", level);
+    getFancyContext().setAllFancyLoggers(*level_to_use);
+  }
+
+  return true;
+}
+
+bool LogsHandler::changeLogLevels(
+    const absl::flat_hash_map<absl::string_view, absl::string_view>& changes) {
+  bool ret = true;
+  for (auto [name, level] : changes) {
+    ret = changeLogLevelByName(name, level) && ret;
```

Comment:
This is still n^2, as changeLogLevelByName does a loop to find it.

You need to pass the map into changeLogLevelByName, and just do one loop, doing map.find().

Also, please validate all names first, reporting any errors in one summary line. Only do mutations if the command is error-free.